### PR TITLE
[perf, model] feat: pack GQA Ulysses all-to-all

### PR DESCRIPTION
**Pull request by 0z5a**
### What does this PR do?

Extend packed Ulysses QKV communication to GQA/MQA-style layouts where Q and KV
head counts differ. Per-destination Q/K/V chunks are flattened into one payload,
redistributed with one autograd-capable `all_to_all_single`, and reconstructed
with the original tensor shapes.

The variable-payload path is selected only when tensor rank, dtype, device, and
all non-head dimensions match, the sequence and head dimensions differ, and
each head count is divisible by the sequence-parallel world size. Other layouts
retain the legacy three-collective fallback. GLM4V and Qwen2-VL also avoid
expanding KV heads when their native head counts are already divisible by the
SP world size.

This is PR 3 of 3 and depends on [PR 2](https://github.com/verl-project/verl/pull/7418).
This is a stacked PR, so its diff includes PR 1 and PR 2 until those dependencies merge.

### Checklist Before Starting

- [x] Searched open PRs for similar work:
  https://github.com/verl-project/verl/pulls?q=is%3Apr+is%3Aopen+ulysses+qkv
  Existing results concern other Ulysses/model support and do not implement the
  variable-payload packed algorithm used here.
- [x] Title follows `[{modules}] {type}: {description}`:
  `[perf, model] feat: pack GQA Ulysses all-to-all`.

### Test

Run on 2 x NVIDIA A100-SXM4-40GB:

```bash
NCCL_P2P_DISABLE=1 NCCL_SHM_DISABLE=1 \
torchrun --standalone --nproc-per-node=2 -m pytest -q \
  tests/special_distributed/test_packed_qkv_ulysses.py
```

Result on A100 before removing the projection-only tests: `21 passed` on each
rank. The current PR retains 14 of those tests. GQA forward/backward matches
three legacy collectives exactly for FP32, BF16, FP8 e4m3, and FP8 e5m2. Tests
also assert one `all_to_all_single` call and verify the incompatible-layout
fallback.

A BF16 mini-Llama GQA inference comparison produced identical logits
(`max_abs_error=0`) and exercised two variable-payload calls.

The current three-PR series head was independently tested on 2 x NVIDIA H200 NVL
with PyTorch 2.11.0+cu128 and default NCCL settings: `14 passed` on each rank in
9.11 seconds.

```bash
pre-commit run --all-files --show-diff-on-failure --color=always
```

Result: every hook passed, including Ruff, Ruff format, mypy, config generation,
license, device API, repository structure, and compileall checks.

Primary microbenchmark: 2 x A100, world size 2, local sequence 256, Q/KV heads
16/4, head dimension 128, 20 warmups, and 100 iterations.

| dtype | 3 x A2A | variable packed | speedup |
|---|---:|---:|---:|
| FP32 | 1.660 ms | 1.407 ms | 1.180x |
| BF16 | 0.924 ms | 0.839 ms | 1.101x |
| FP8 e4m3 | 0.524 ms | 0.473 ms | 1.107x |
| FP8 e5m2 | 0.533 ms | 0.454 ms | 1.175x |

Supplementary H200 result from the new three-PR stack at local sequence 256:
paired-speedup median and range across three runs, each with 100 warmups and
1000 iterations.

| dtype | median speedup | observed range |
|---|---:|---:|
| FP32 | 1.077x | 1.009x - 1.083x |
| BF16 | 0.981x | 0.915x - 1.132x |
| FP8 e4m3 | 1.049x | 0.967x - 1.289x |
| FP8 e5m2 | 1.015x | 0.968x - 1.270x |

The same H200 host had unrelated GPU processes earlier in the session, but
`nvidia-smi` reported zero allocated memory immediately after these new runs.
The cross-NUMA `SYS` topology and remaining run-to-run variation make this a
supplementary result; the less-variable A100 sample is the primary comparison.

### API and Usage Example

```python
from verl.utils.ulysses import gather_qkv_seq_scatter_heads

# q may have more heads than k/v; the helper chooses the variable-payload path
# when each head count is divisible by the SP world size.
q, k, v = gather_qkv_seq_scatter_heads(q, k, v, seq_dim=2, head_dim=1)
```

No new model configuration is required. Existing callers use the same helper;
its dispatch now recognizes safe GQA/MQA layouts.

### Design & Code Changes

- Build one per-destination flattened payload containing that destination's Q,
  K, and V head chunks.
- Exchange variable split sizes with one autograd-capable
  `all_to_all_single`, then reconstruct the original logical tensors.
- Retain the legacy three-collective path for incompatible layouts.
- Avoid unnecessary KV-head expansion in GLM4V and Qwen2-VL when native KV
  heads already satisfy sequence-parallel divisibility.
- Extend distributed forward/backward, call-count, fallback, FP32, BF16, FP8,
  mini-Llama inference, and microbenchmark coverage.

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md).
- [x] Applied the required full pre-commit command; all hooks passed.
- [x] Documentation is covered by utility docstrings and the usage example
  above; there is no new end-user configuration or CLI surface.
- [x] The model CI workflow added in PR 1 runs this PR's expanded distributed
  test file.
- [ ] Human submitter has reviewed every changed line and can defend the change
  end-to-end.
- [ ] After the PR is opened and ready, request CI in the verl `ci-request`
  Slack channel or Feishu group.
- [x] Recipe submodule update is not applicable.

This change was developed with OpenAI Codex assistance. The test, inference, and
benchmark commands and their results are disclosed above.

### L20 GQA scaling boundary (2026-09-03)

The full L20 topology/scaling matrix and Nsight evidence are recorded in
#7417. With Q/KV heads 48/12 and BF16, GQA speedup at WS3 was 1.019×, 1.255×,
1.003×, and 0.961× for local sequence 256/2K/8K/32K. WS2 was 0.936×, 0.663×,
0.975×, and 0.931×. FP8 e4m3 WS3 was 0.872× at 256 and 0.990× at 2K.

Distributed correctness remained exact at WS2/WS3 for forward, backward,
FP32/BF16/FP8, and incompatible fallback (`14 passed` per rank). Nsight on
WS3/BF16/2K confirmed one NCCL SendRecv kernel per rank instead of three, but
aggregate NCCL duration was 29.940ms versus 28.856ms for legacy. Thus the
variable-payload path meets its launch-count contract while exposing a real
topology/workload crossover rather than a universal speedup.

### CODE REVIEW

File: verl/utils/ulysses.py [hunk 1]
```diff
@@ -87,6 +87,199 @@ def gather_seq_scatter_heads(
...
+    equal_layout = (
+        query.shape == key.shape == value.shape
+        and query.dtype == key.dtype == value.dtype
+        and query.device == key.device == value.device
+        and seq_dim != 0
+        and head_dim != 0
+    )
+    if equal_layout:
+        batch_sizes = (query.size(0), key.size(0), value.size(0))
+        packed_qkv = torch.cat((query, key, value), dim=0)
```
Comment (wuxibin89) [inline verl/utils/ulysses.py:160]: Pack qkv into one all-to-all increase peak memory overhead and introduce additional cuda memory copy overhead. For long-context the overhead will surpass the gain reducing 2 nccl kernel launch.
